### ISSUE TITLE
To-Do List isn't saving the sort order [$56 awarded]

### ISSUE THREAD

**Issue by bingbangzoom** [HabitRPG/habitica#2606]
Every time I rearrange my To-Dos, and refresh the page or come back to it, the order is messed up again. How can I fix this? 

_UPDATE by admin, October 2015:_

**This bug has been mostly fixed. Please see https://github.com/HabitRPG/habitrpg/issues/6040 for a description of a remaining, minor To-Do sorting bug.**

_edit by admin:_

What to do if you have this problem:

If you experience any problems with sorting To-Dos, delete your completed To-Dos. Export them first if you want to keep a permanent record of them. You might need to do this every few days. To learn how to delete and export To-Dos, refer to the "Clean up" section in [To-Dos](http://habitrpg.wikia.com/wiki/To-Dos).

You don't need to comment here if you notice this problem because it happens to everyone.

_end of edit by admin_
##

**Comment by theglus** [comment]
I seem to be having the same issue. It just started happening on 1/24/14, it was working perfectly fine on 1/23/14 (at least for me). I've experienced the issue in both Chrome (32.0.1700.77) and Firefox (26). I'm running OSX 10.9.1. It sounds like it might be related #2180 and #1616

**Comment by MrSmearkase** [comment]
Howdydo, I too am experiencing this issue. Every refresh or sync operation restores the previous order from before my sort. I am running Chromium on Ubuntu 13.10.

**Comment by mistergid** [comment]
I'm also having trouble with this. The minute I refresh, they rescramble. Firefox 26 on OSX 10.9.1, Chrome 32.0.1700.76 on Windows 7.

**Comment by daraxlaine** [comment]
I am also having this problem. I have this problem on my home computer (windows 8, latest google chrome) and my work computer (windows 7, latest google chrome: Version 32.0.1700.102 m)

**Comment by lefnire** [comment]
Definitely has to do with archived todos. The full to-do array is preened before sending to client. The sort `to` and `from` indicies are incorrect when handled on the server, which works with the full array instead of the preened. It's official, we're gonna need to split `user.todos` into `user.todos, user.completed`. Too many things have been prevented by not having that difference, methinks this tick is the last straw

**Comment by lefnire** [comment]
and this proves my theory: if you clear your completed to-dos, the problem is fixed. The reason is clearing completeds also clears archived to-dos, so then the client matches the server. Clear completed, refresh, sort, √ ('course, that's not a real fix - but good workaround for now)

**Comment by theglus** [comment]
@lefnire Thanks for the tip. I just cleared my completeds, and sure enough now I can successfully order my To Dos. That's a sufficient fix for me for now :)

**Comment by lorian** [comment]
The biggest issue with the "clear completeds" fix is that if you're a member of a challenge with to-dos, they can't be cleared until the challenge is complete :/

**Comment by lefnire** [comment]
That's fine, chal todos aren't archived, so client will still match server

**Comment by lorian** [comment]
...huh. Then why are my to-dos still rearranging when all I have are
challenge to-dos completed? I'll have to keep a closer eye on this.

**Comment by ghost** [comment]
Am I understanding correctly that to fix this, you need to clear To Dos which also erases your To Do history? Is it right to interpret this to mean that the 'retain full history' perk for subscribers doesn't actually happen if To Dos are cleared?

**Comment by lefnire** [comment]
Retain full history is about history graphs. When you click the graph icon
on your tasks, you get a chart of your task histories. Same for overall
experience (which tracks your general progress). These histories are
truncated for most users for performance and database optimization. Retain
full history doesn't pertain to archived to-dos. If people want that, I can
work towards adding that to the mix

**Comment by lorian** [comment]
I would love to be able to archive all my completed to-dos, but only if they were marked with completion date. It would help me look up that, say, I last washed the sheets two weeks ago.

**Comment by ghost** [comment]
Ah, I misunderstood. Sorry. Completed To Dos that could be reviewed would be nice (I could get rid of most of my personal Trello board). At most I would need a month worth of history. I can't see needing to look back further than that (for myself).

**Comment by caveman1337** [comment]
Issue was fixed after deleting Completed To-dos, despite nothing showing there. I'm surprised user.todos and user.completed were ever in the same array in the first place.

**Comment by colegleason** [comment]
They are in the same array because, well, they are essentially the same thing. An uncompleted todo and a completed only have one difference: a simple boolean flag change. That's much easier (conceptually, performance-wise, etc) to be represented as a `todo.complete = true` instead of moving it to another array,

**Comment by caveman1337** [comment]
That's assuming you write once and don't care about the ordering. The problem with storing them in the same array is that when you sort them around, it takes a lot more performance and can cause more issues if the completed ones are hidden from the list. If you put them in two arrays, you don't have to sort through a potentially huge list every time you want to change things on your to-do. It is much easier and efficient to move completed tasks to a separate array and sort a smaller list of current tasks than it is to sort a single array containing both. This is also better visually as the completed and current tasks aren't shown together; which complicates the sorting function even more when they are together in a single array.

**Comment by lefnire** [comment]
@caveman1337 yes yes, agreed - and we'll move to separate arrays the moment we can.

**Comment by RainbowMagicMarker** [comment]
When trying to sort my to-do list I have a fairly large checklist I want to keep at the bottom, however every time I try to sort my To-do's with deadlines by date and keep anything that doesn't have a deadline below those, however when I try to move one to it's new location, that one as well as the two it would have been neighbors with go down to the bottom of my list. They're in the right order but now they're beneath my big checklist. I cannot bring them back up above, they simply revert to the bottom when I release the mouse button, and they cannot be moved until I try to move something else and then those three go to the bottom instead.

Windows 7 on Firefox 24.0 and I have already tried clearing my completed to-dos, it did not fix this issue.

**Comment by lorian** [comment]
@RainbowMagicMarker I had the exact same bug (with 2 to-dos suddenly deciding to only move to the very bottom of my to-do list, no matter where I tried to put them), but it was fixed when I reloaded the page. I'm not sure what actually cured it, however, because the reload also caused me to get the new UI, so I'm sure lots of weird things were going on behind the scenes during that reload.

**Comment by LittleRabbitBeta** [comment]
My to-dos have been jumping arounds somewhat randomly.

**Comment by caveman1337** [comment]
I suspect the to-dos "jumping" has a bit to do with the items being displayed in their right order, but the drag feature associates the box as whatever position in the array that it would be, even if it corresponds to a hidden completed to-do. 

So you pick the 3rd box displayed, it thinks you grabbed the 3rd item in the array, and positions it accordingly. This causes problems because the hidden completed tasks are stored in their original order in the same array. Example: The 2nd item in the array is a completed task, and the rest of the items are to-dos. If you move the 3rd item to the bottom, then it will move the 3rd item in the array (which is the 2nd item that is visible to you) to the bottom instead. This is my speculation based on trial and error, rather than looking at the code, so I may be a bit off.

**Comment by storian** [comment]
Hi, I am currently experiencing this problem.  It is still corrected by removing my completed items.  All of my completed items contained checklists, in case that matters.

**Comment by deilann** [comment]
@storian This issue is still open?

**Comment by storian** [comment]
Yes, I have been having this problem this morning.

**Comment by storian** [comment]
(It looks like it actually is still open above.  Lots of related issues are closed but this one appears not closed.)

**Comment by deilann** [comment]
Yes, that's what I meant. The related issues are closed because they're
duplicates. We only need one ticket on this issue! :) But this one is
definitely still open.

**Comment by deilann** [comment]
Might just be a server issue. Are you also making sure to let the blue bar
move all the way before moving some things?

**Comment by brightbold** [comment]
Wow Ryan you replied to that super fast, before I deleted my post. :) I didn't have any completed To Dos listed, but I cleared them anyway and that solved the problem. I had failed to read the notice about archived To Dos, so I'm guessing that invisible archived to dos were causing the problem.

**Comment by deilann** [comment]
I'm magic. :)

**Comment by notapooka** [comment]
I'm now experiencing this issue even after clearing my completed To Dos. I've cleared about ten times, and things still reorganize themselves into a completely different order. It's incredibly frustrating, because the task I REALLY need to get done are all getting punted to the middle of my list, where I easily lose track of them.

**Comment by lorian** [comment]
@notapooka I had that situation once, and closing and re-opening the browser fixed it.

**Comment by deilann** [comment]
It may not be saving to the server, which is a different boat entirely, and
can affect all tasks.

When moving tasks, always make sure that the blue bar finishes entirely
before touching anything else. Then, do a refresh to see if it "sticks."

If all else fails, go for a tag. Tag things as priority, and you'll be able
to find them easily.

**Comment by notapooka** [comment]
@lorian Thank you! Closing and re-opening my browser seems to have done the job.

**Comment by cesarioviola** [comment]
Closing and reopening is not doing it for me; neither is clearing my completeds. Other ideas?

**Comment by deilann** [comment]
Clearing your completed should definitely be it. Are you sure you're
waiting for the blue line to finish every time you move something?

**Comment by cesarioviola** [comment]
I did. It happened twice more after I had cleared the completeds, but it's stopped now, so I don't know. Oh well, at least it's stopped. Thanks for looking into it.
On Thursday, April 10, 2014 10:26 PM, deilann notifications@github.com wrote:

Clearing your completed should definitely be it. Are you sure you're
waiting for the blue line to finish every time you move something?

**Comment by RankkaApina** [comment]
For a long time it worked for me when I cleared the completed. Now for 2 days it's not working. There's no way I can arrange my to-dos (which is quite a long list so sorting would be helpful...)

**Comment by deilann** [comment]
Are you making sure to wait for the blue bar to finish every time you move
something? That will also interrupt the sort process.

**Comment by RankkaApina** [comment]
Yes, I even checked it as I saw it suggested here. The thing is, this has worked before, but since last night everything scrambles up when the page is refreshed.

**Comment by lefnire** [comment]
In the next few months (after I've shaped up mobile a bunch), I'll be working on a fix for this. It's a larger issue, requiring moving task lists to objects, and separating `user.todos` into `user.todos` & `user.completed`. This larger project will solve many of the challenges bugs, the VersionError, and much more.

**Comment by deilann** [comment]
Weird. Do you have any tags on? Even if it doesn't look like it, try
clearing them.

**Comment by RankkaApina** [comment]
Tried clearing the tags, no help, I even cleared the cookies just in case :( but good to know someone's working on this. I'll try to get by. Maybe it will start working as mysteriously as it stopped working...

**Comment by deilann** [comment]
Are you on beta or prod?

@lefnire just mentioned that beta's being futzy.

**Comment by RankkaApina** [comment]
I'm not sure, I'm quite new... How can I find out?

**Comment by deilann** [comment]
If you're not specifically accessing beta.habitrpg.com you are on the prod
server. :) So probably prod.

**Comment by lorian** [comment]
I'm REALLY looking forward to a fix for this; just like @RankkaApina , my to-dos rearrange themselves despite having cleared my tags, completeds, waiting for the blue bar, and even closing/opening the browser entirely (definitely not on beta, either). It was working fine for months until this afternoon; now whenever I refresh, things move back to quasi-random locations (near, but not exactly, where they were before I moved them).

**Comment by deilann** [comment]
Until this afternoon. That's interesting -- I wonder if it has to do with
the archiving thing that was merged.

**Comment by brightbold** [comment]
I agree with @lorian that the behavior is now different. Clearing completed (and waiting for the blue bar) always worked for me before, but now it doesn't. The task either moves part way to where I want it but not all the way, or jumps back from whence it came. (I even tried overshooting the mark, hoping it would end up where I wanted it, but no dice.) So there may be something too your theory that the archiving thing changed something.

**Comment by deilann** [comment]
@lefnire Do you think that's what did it? Maybe a revert is in order, at
least until things are changed so it's no longer even a problem?

**Comment by lorian** [comment]
Yep, my observed behavior is the same as @brightbold , and the issue is still ongoing despite a browser close and reboot. This is by far the longest I've ever had the to-do list order problem, so I think you may be right that something has actually changed.

**Comment by RankkaApina** [comment]
I'd love a revert until this is fixed. I don't mind clearing the completed, but not being able to order my to-dos is a problem since it's a long list... At the moment I need to keep another list on my computer in the order I want and then find it in habitrpg...

Also before when I cleared the completed the page updated with removing them from the completed list. Now I have to refresh the page for that to happen. This doesn't bother me, but I thought it might be related...

**Comment by lefnire** [comment]
@jbutz looks like the todo archiving stuff has caused some issues with sorting. Do you have any proposed quick-fixes on this, or shall we revert the commits for the time being?

**Comment by jbutz** [comment]
@lefnire  If it started 3 months ago I don't think my commits would have caused it, but if reverting helps then go for it. I won't have time to look for a fix for a few days.

**Comment by deilann** [comment]
The issue started a long time ago, but the workaround that made it possible
to handle broke around the time that your code went live.

**Comment by brightbold** [comment]
Bump? I would also love a temporary revert. Not being able to pull high-priority tasks to the top of the list is hurting productivity.

**Comment by lefnire** [comment]
I'll revert tonight if I can't get a proper fix in place by then

**Comment by deilann** [comment]
Thanks @lefnire

**Comment by brightbold** [comment]
Thanks - appreciate it!

**Comment by lefnire** [comment]
Feature reverted (https://github.com/HabitRPG/habitrpg/commit/e77f667a98db678668ce70a964d9be35a8f8d965 & https://github.com/HabitRPG/habitrpg-shared/commit/519154dc8932ce3f442140b1e718d45ff7ac0224)

**Comment by lorian** [comment]
After a shift-reload of the page, I can now successfully move to-dos and have them stick in their new location! Also, clearing my completed to-dos causes them to visibly clear. Thank you very much for the revert!

**Comment by Cassolotl** [comment]
I'm not sure how helpful or relevant this is but I just wanted to pop in and say that I really really would like to be able to reorder to-dos without clearing completed to-dos. Putting the stuff I want to get done today at the top would help me so much. But also, I don't like clearing my completeds. I find it helpful to see what I've done in the last few days, to get an idea of how productive I've been and to feel a bit better about the stuff I haven't done yet. ("Boo, I am rubbish, look at all these red to-dos! But, hang on, I'm actually pretty awesome because look at all these completeds.") The weirdness of to-do reordering is putting me off using HabitRPG, because to-dos is where I get most of my XP.

Is this issue basically going to be around forever? It sounds like a massive pain to fix, and like it would require a lot of rejiggering, and that might be more hassle than the alternatives. Does adding to the bounty help?

Sorry if these questions and comments are not very helpful. I just don't know who/where else to ask.

**Comment by lefnire** [comment]
I'm working on the fix at https://github.com/habitrpg/habitrpg/tree/tasks-obj . It will solve this issue, challenge-closing issues (for large challenges), and VersionError issue. It's high/medium prio, but it will take a long time. Feel free to help out ;)

**Comment by Cassolotl** [comment]
How exciting! I lack all relevant skills so I'm not sure what I can do except cheerlead... _\o/_ _\o/_ _\o/_ _\o/_ woooo go lefnire!

Also, thank you very much. :)

**Comment by Verabird** [comment]
I would like to help with this. @lefnire if there's anything you think I should know, my ears (eyes?) are open. Checking out your repository now to try to get oriented.

**Comment by lefnire** [comment]
@Verabird thanks for that! Actually, TBH this might be one I have to tackle solo. I'm working a branch `tasks-obj` ([web](https://github.com/habitrpg/habitrpg/tree/tasks-obj), [shared](https://github.com/habitrpg/habitrpg-shared/tree/tasks-obj)). This branch is the fix for many issues, including this one. However, due to rebasing craziness I've been doing interactive rebases & force-pushes; meaning collab on this branch will probably be too difficult. 

Once I get it to a stable point later, can I ping you for help in testing / refactoring / edge-case-hunting / etc?

**Comment by Verabird** [comment]
Sure, ping me for what you need when you need it

**Comment by SabreCat** [comment]
@lefnire: The "push to top" feature gave me an idea for how to mitigate this until a permanent solution can be implemented. If I understand correctly, hidden completed To-Dos are disrupting the sort order--you move an active To-Do, and it tries to snap to a position before or behind something you can't see. What if, when you complete a To-Do, it does a "push to _end_" behind the scenes? Then you wouldn't have any ordering conflicts except when moving something to the very bottom of your list.

**Comment by Ulugbeg** [comment]
Sounds cool, downside would be that if you undelete it, it does not go back to its previous location but to the bottom, too, but I'm wiling to live with that in a heartbeat :) Good suggestion!

**Comment by SabreCat** [comment]
@Ulugbeg: Hmm, good point. Unchecking would re-introduce sorting confusion too... unless uncheck did a push to top? xD A little kludgy, but might be sufficient for an interim fix.

**Comment by Alys** [comment]
We could sell it as a new feature. :) "If you accidentally tick off the wrong to-do, you'll always find it at the bottom of the completed to-dos list, just above the tab that you just used to see your completed to-dos."

**Comment by cesarioviola** [comment]
How do I get unsubscribed from this conversation?

On Tuesday, September 2, 2014 9:11 PM, Alice Harris notifications@github.com wrote:

Is anyone working on SabreCat's interesting  "push to bottom" workaround? If not, I'll give it a go over the next few days.

**Comment by deilann** [comment]
Go to the thread on Github and hit the unsubscribe button.

**Comment by Alys** [comment]
These two PRs implement SabreCat's suggested workaround (https://github.com/HabitRPG/habitrpg/issues/2606#issuecomment-53584917):
https://github.com/HabitRPG/habitrpg/pull/4007
https://github.com/HabitRPG/habitrpg-shared/pull/304

**Comment by stimulus2da** [comment]
Not that this comment will be very helpful, but add me to the list of people having this issue, on at least two computers running Windows 7 and Chrome (currently 37.0.2062.120 m). I'm new to GitHub, so if there is a better way for me to say that I'm having the issue without clogging up the comments, please let me know. Thanks!

**Comment by Ulugbeg** [comment]
If I read everything correctly, this should be 'fixed' now - since about 4 days. However, if you still have old completed tasks, I think things might still go wrong.

To fix that:

A. Clear your old completed tasks. New completed tasks should no longer cause sorting issues with the uncompleted tasks.

OR

B. 
1. Complete 1 task. 
2. Go to uncompleted tasks
3. Put all the tasks below the task you completed in step 1.

I THINK both workarounds should sort (hah.) the problem.

Edit: note that I haven't looked at code, but I think the proposed workaround has been pushed to the live site. The proposed workaround has conceptually the problem that it only works by preventing the problem for completed todo's, not fixing it for completed todo's which are already causing this problem.

**Comment by stimulus2da** [comment]
I did option A and it worked perfectly. Thanks for laying that out for me, I guess my coffee hasn't kicked in yet!

Cheers

**Comment by Ulugbeg** [comment]
Good, glad it worked. Option A is by far the easiest, but option B is for people who would like to keep their old todo's and don't mind doing a bit of extra work to get it to work :)

**Comment by DrTall** [comment]
I joined the site a few weeks ago and I have noticed that only recently my To-Do list order is not saving drags and drops. It seemed to work just fine for awhile, but as of maybe yesterday I am not able to make changes to the ordering. Is anyone else seeing this?

**Comment by Alys** [comment]
@DrTall yes, that's what this issue is about, so everyone experiences it. The workaround of deleting your completed To-Dos will work.

This reminds me that I've been meaning to comment on [my change](https://github.com/HabitRPG/habitrpg/issues/2606#issuecomment-54935416) for [SabreCat's suggested workaround](https://github.com/HabitRPG/habitrpg/issues/2606#issuecomment-53584917). It doesn't seem to have made a difference, does it. :( Which is a bit weird because I thought SabreCat's logic was impeccable. Maybe there's something else going on that we don't understand? Or maybe kittehs are just bad at logic. ;)

**Comment by stimulus2da** [comment]
Seeing DrTall's comment reminded me to follow up on this:

While deleting my todos fixed the issue for me for a while, I've noticed that every once in a while I still get this issue popping up. I thought that it may be due to old tasks FINALLY being completed and then impeding the movement, however, it has happened solely with tasks created since the work around.

Deleting my old tasks re-fixes the issue and I can move tasks with ease again. It's not a huge issue for me since I don't care if I have to delete my old tasks, but it does seem like there's something else going on.

**Comment by DrTall** [comment]
Oh I mis-read the github status. I thought this issue was marked as Closed. As it is still Open, I will just do the work around. Thanks.

**Comment by pydsigner** [comment]
@lefnire Are you still working on this?

**Comment by lefnire** [comment]
@pydsigner I'm working on a big migration (from tasks stored as arrays to a separate collection) which will fix this issue.

**Comment by cdepillabout** [comment]
My todo order constantly gets messed up, even when I have no completed tasks.  I hope this issue gets fixed soon.

However, I don't know if I'll be able to live with a HabitRPG where everything works correctly all the time :-\

**Comment by dcecile** [comment]
@lefnire I know you're working on a working on a bigger fix to this issue. How about an interim solution? I've created a pull request, #6018.

It doesn't support the case when an "archived" todo is before the 'from' index, but it does support the case when an "archived" todo is before the 'from' index (e.g. moving a new task to the top of your todo list).

What do you think?

**Comment by Alys** [comment]
@dcecile : lefnire is no longer working on this, but other high-tier devs are. They're making large-scale changes to how tasks are stored, and the end result of that might mean that this bug is no longer relevant. But this does not mean that we aren't interested in an interim solution. :)

I haven't had a chance to test your code yet or even examine it thoroughly (I'm at work and on a brief break), but at first glance your idea looks really clever! We'll definitely look into your PR. Thank you!

**Comment by dcecile** [comment]
@Alys : You're welcome!

Take a look at the PR, give it a spin, let me know what you think. :)

**Comment by brightbold** [comment]
@dcecile I am so excited about your fix! This issue has been driving me particularly crazy recently, so it's really exciting to see that a solution may be imminent and not held up by a large-scale refactoring. Thanks!

@Alys I don't know how this works but I'd vote for giving Dan the bounty, even though it sounds like this solution isn't 100% perfect. As one of the bounty contributors on this issue, I'm much happier to have an imperfect fix now (that might sometimes require me to move a task twice to get it to stick) than to continue having to decide whether to delete my task history or have all my tasks out of order while I wait for a big code refactoring to be completed. 

Can't wait for this to be deployed!

**Comment by Alys** [comment]
This issue has been fixed in almost every case by @dcecile  - see https://github.com/HabitRPG/habitrpg/pull/6018#issuecomment-147582791 for details. I have asked dcecile to claim the bounty.
